File: android/core/src/main/kotlin/org/hdhropen/kit/viewmodels/AppEnvironment.kt
package org.hdhropen.kit.viewmodels

import android.content.Context
import androidx.media3.common.util.UnstableApi
import org.hdhropen.kit.networking.APIClient
import org.hdhropen.kit.networking.AuthManager
import org.hdhropen.kit.networking.ServerDiscovery
import org.hdhropen.kit.networking.WatchSessionManager
import org.hdhropen.kit.playback.PlaybackPreferences
import org.hdhropen.kit.playback.PlayerEngine
import org.hdhropen.kit.theme.ThemePreferences

@UnstableApi
class AppEnvironment(
    val context: Context,
    defaultServerURL: String = "http://127.0.0.1:8000"
) {
    val serverDiscovery = ServerDiscovery(context, defaultServerURL)
    val apiClient = APIClient(baseURL = serverDiscovery.serverURLString.value)
    val authManager = AuthManager(apiClient, context)
    val watchSessionManager = WatchSessionManager(apiClient)
    val playerEngine = PlayerEngine(context)
    val playbackPreferences = PlaybackPreferences(context)
    val themePreferences = ThemePreferences(context)

    val guideViewModel = GuideViewModel(apiClient)
    val recordingsViewModel = RecordingsViewModel(apiClient)
    val playerViewModel = PlayerViewModel(apiClient, watchSessionManager, playerEngine, playbackPreferences = playbackPreferences)
    val tunerViewModel = TunerViewModel(apiClient)
    val settingsViewModel = SettingsViewModel(apiClient)
    val authViewModel = AuthViewModel(authManager)

    fun updateServerURL(url: String) {
        serverDiscovery.setServerURL(url)
        apiClient.baseURL = url
    }
}

File: apple/HDHROpenKit/Sources/HDHROpenKit/ViewModels/AppEnvironment.swift
import Foundation
import SwiftUI

@MainActor
public final class AppEnvironment: ObservableObject {
    public let apiClient: APIClient
    public let authManager: AuthManager
    public let serverDiscovery: ServerDiscovery
    public let watchSessionManager: WatchSessionManager

    public let guideViewModel: GuideViewModel
    public let recordingsViewModel: RecordingsViewModel
    public let tunerViewModel: TunerViewModel
    public let settingsViewModel: SettingsViewModel
    public let authViewModel: AuthViewModel
    public let playerViewModel: PlayerViewModel
    public let multiPlayerViewModel: MultiPlayerViewModel
    public let themeManager: ThemeManager

    public init(defaultURLString: String = "http://127.0.0.1:8000") {
        let discovery = ServerDiscovery(defaultURL: defaultURLString)
        let initialURL = discovery.currentServerURL ?? URL(string: defaultURLString)!
        let client = APIClient(baseURL: initialURL)
        let auth = AuthManager(apiClient: client)
        let watch = WatchSessionManager(apiClient: client)

        serverDiscovery = discovery
        apiClient = client
        authManager = auth
        watchSessionManager = watch

        guideViewModel = GuideViewModel(apiClient: client, watchSessionManager: watch)
        recordingsViewModel = RecordingsViewModel(apiClient: client)
        tunerViewModel = TunerViewModel(apiClient: client)
        settingsViewModel = SettingsViewModel(apiClient: client, serverDiscovery: discovery)
        authViewModel = AuthViewModel(authManager: auth)
        playerViewModel = PlayerViewModel(apiClient: client, watchSessionManager: watch)
        multiPlayerViewModel = MultiPlayerViewModel(apiClient: client, watchSessionManager: watch)
        themeManager = ThemeManager()
    }

    public func setServerURL(_ url: URL) async {
        serverDiscovery.serverURLString = url.absoluteString
        await apiClient.setBaseURL(url)
        await authManager.restoreSession()

        async let guide: Void = guideViewModel.loadData()
        async let recordings: Void = recordingsViewModel.loadData()
        async let tuners: Void = tunerViewModel.loadData()
        async let settings: Void = settingsViewModel.loadData()
        async let profiles: Void = authManager.fetchProfiles()
        _ = await (guide, recordings, tuners, settings, profiles)
    }
}
